flocker/node/test/test_loop.py

```diff
@@ -304,62 +268,118 @@ def successful_amp_client(self, local_states):
 
     def test_convergence_done_notify(self):
         """
-        A FSM doing convergence that gets a discovery result send the
+        A FSM doing convergence that gets a discovery result, sends the
         discovered state to the control service using the last received
         client.
         """
-        local_state = object()
+        local_state = NodeState(hostname=b'192.0.2.123')
         client = self.successful_amp_client([local_state])
-        action = ControllableAction(Deferred())
+        action = ControllableAction(result=Deferred())
         deployer = ControllableDeployer([succeed(local_state)], [action])
         loop = build_convergence_loop_fsm(Clock(), deployer)
-        loop.receive(_ClientStatusUpdate(client=client,
-                                         configuration=object(),
-                                         state=object()))
+        loop.receive(
+            _ClientStatusUpdate(
+                client=client,
+                configuration=Deployment(
+                    nodes=frozenset([local_state.to_node()])
+                ),
+                state=Deployment(
+                    nodes=frozenset([local_state.to_node()])
+                )
+            )
+        )
         self.assertEqual(client.calls, [(NodeStateCommand,
                                          dict(node_state=local_state))])
 
+    def test_convergence_done_update_local_state(self):
+        """
+        An FSM doing convergence that gets a discovery result supplies an
+        updated ``cluster_state`` to ``calculate_necessary_state_changes``.
+        """
+        local_node_hostname = u'192.0.2.123'
+        # Control service reports that this node has no manifestations.
+        received_node = Node(hostname=local_node_hostname)
+        received_cluster_state = Deployment(
+            nodes=frozenset([received_node])
+        )
+        discovered_manifestation = Manifestation(
+            dataset=Dataset(dataset_id=uuid4()),
+            primary=True
+        )
+        local_node_state = NodeState(
+            hostname=local_node_hostname,
+            manifestations=[discovered_manifestation]
+        )
+        client = self.successful_amp_client([local_node_state])
+        action = ControllableAction(result=Deferred())
+        deployer = ControllableDeployer([succeed(local_node_state)], [action])
+
+        fsm = build_convergence_loop_fsm(Clock(), deployer)
+        fsm.receive(_ClientStatusUpdate(client=client,
+                                        configuration=object(),
```

Comment:
We decided to stop using wrong object types in tests, this should be a `Deployment` instance.